File: fireplace-app/src/main/java/io/github/bric3/fireplace/charts/Range.java
package io.github.bric3.fireplace.charts;

import org.jetbrains.annotations.NotNull;

import java.util.Objects;

/**
 * Represents a range of values between min and max.
 */
public record Range<T extends Number>(@NotNull T min, @NotNull T max) {
    /**
     * Creates a new range.
     *
     * @param min the minimum value.
     * @param max the maximum value.
     */
    public Range {
        check(min, max);
    }

    private static <T extends Number> void check(T min, T max) {
        if (min == null || max == null) {
            throw new IllegalArgumentException("Null 'min' or 'max' argument.");
        }
        if (min.getClass() != max.getClass()) {
            throw new IllegalArgumentException("min and max must be of the same type, min is " + min.getClass() + ", max is " + max.getClass());
        }
        if (min instanceof Double && min.doubleValue() > max.doubleValue()) {
            throw new IllegalArgumentException("min must be less than max, min is " + min + ", max is " + max);
        }
        if (min instanceof Long && min.longValue() > max.longValue()) {
            throw new IllegalArgumentException("min must be less than max, min is " + min + ", max is " + max);
        }
        if (min instanceof Integer && min.intValue() > max.intValue()) {
            throw new IllegalArgumentException("min must be less than max, min is " + min + ", max is " + max);
        }
    }

    public boolean isZeroLength() {
        return min.equals(max);
    }

    /**
     * Returns a range based on this range but extended (if necessary)
     * to include the specified value.
     *
     * @param v the value.
     * @return The range.
     */
    public Range<T> include(T v) {
        if (v instanceof Double && !Double.isFinite(v.doubleValue())) {
            throw new IllegalArgumentException("only finite values permitted");
        }
        if (compare(v, min) < 0) {
            return new Range<>(v, max);
        }
        if (compare(v, max) > 0) {
            return new Range<>(min, v);
        }
        return this;
    }

    /**
     * Calculates a fractional value indicating where v lies along the
     * range.  This will return 0.0 if the range has zero length.
     *
     * @param v the value.
     * @return The fraction.
     */
    public double calcFraction(T v) {
        if (compare(max, min) > 0) {
            if (min instanceof Double) {
                return (v.doubleValue() - min.doubleValue()) / (max.doubleValue() - min.doubleValue());
            }
            if (min instanceof Long) {
                return (double) (v.longValue() - min.longValue()) / (double) (max.longValue() - min.longValue());
            }
            if (min instanceof Integer) {
                return (double) (v.intValue() - min.intValue()) / (double) (max.intValue() - min.intValue());
            }
            throw new IllegalArgumentException("Unsupported type " + min.getClass());
        }
        return 0.0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        var range = (Range<?>) o;
        if (min.getClass() != range.min.getClass()) {
            return false;
        }
        return compare(range.min, min) == 0 && compare(range.max, max) == 0;
    }

    /**
     * compare two {@code Number} of the same type.
     *
     * @param a   the first value.
     * @param b   the second value.
     * @param <T> The type of the number.
     * @return the value {@code 0} if {@code d1} is numerically equal to {@code d2}; a value less than
     * {@code 0} if {@code d1} is numerically less than {@code d2}; and a value greater than {@code 0}
     * if {@code d1} is numerically greater than {@code d2}.
     */
    static <T extends Number> int compare(T a, T b) {
        if (a instanceof Double) {
            return Double.compare((Double) a, (Double) b);
        }
        if (a instanceof Long) {
            return Long.compare((Long) a, (Long) b);
        }
        if (a instanceof Integer) {
            return Integer.compare((Integer) a, (Integer) b);
        }
        throw new IllegalArgumentException("Unsupported type " + a.getClass());
    }

    @Override
    public int hashCode() {
        return Objects.hash(min, max);
    }

    @Override
    public String toString() {
        return "[Range: " + min + ", " + max + "]";
    }
}

File: fireplace-app/src/main/kotlin/io/github/bric3/fireplace/charts/Range.kt
/*
 * Fireplace
 *
 * Copyright (c) 2021, Today - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package io.github.bric3.fireplace.charts

import java.util.*

/**
 * Represents a range of values between min and max.
 * 
 * @param min the minimum value.
 * @param max the maximum value.
 */
@JvmRecord
data class Range<T : Number>(val min: T, val max: T) {
    val isZeroLength: Boolean
        get() = min == max

    /**
     * Returns a range based on this range but extended (if necessary)
     * to include the specified value.
     *
     * @param v the value.
     * @return The range.
     */
    fun include(v: T): Range<T> {
        require(!(v is Double && !java.lang.Double.isFinite(v.toDouble()))) { "only finite values permitted" }
        if (compare(v, min) < 0) {
            return Range(v, max)
        }
        if (compare(v, max) > 0) {
            return Range(min, v)
        }
        return this
    }

    /**
     * Calculates a fractional value indicating where [v] lies along the range.
     * This will return 0.0 if the range has zero-length.
     *
     * @param v the value.
     * @return The fraction.
     */
    fun ratioFor(v: T): Double {
        if (compare(max, min) > 0) {
            if (min is Double) {
                return (v.toDouble() - min.toDouble()) / (max.toDouble() - min.toDouble())
            }
            if (min is Long) {
                return (v.toLong() - min.toLong()).toDouble() / (max.toLong() - min.toLong()).toDouble()
            }
            if (min is Int) {
                return (v.toInt() - min.toInt()).toDouble() / (max.toInt() - min.toInt()).toDouble()
            }
            throw IllegalArgumentException("Unsupported type " + min::class.java)
        }
        return 0.0
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val range = other as Range<*>
        if (min.javaClass != range.min.javaClass) {
            return false
        }
        return compare(range.min, min) == 0 && compare(range.max, max) == 0
    }

    override fun hashCode(): Int {
        return Objects.hash(min, max)
    }

    override fun toString(): String {
        return "[Range: $min, $max]"
    }

    init {
        check(min, max)
    }

    companion object {
        private fun <T : Number?> check(min: T?, max: T?) {
            require(!(min == null || max == null)) { "Null 'min' or 'max' argument." }
            require(min.javaClass == max.javaClass) { "min and max must be of the same type, min is " + min.javaClass + ", max is " + max.javaClass }
            require(!(min is Double && min.toDouble() > max.toDouble())) { "min must be less than max, min is $min, max is $max" }
            require(!(min is Long && min.toLong() > max.toLong())) { "min must be less than max, min is $min, max is $max" }
            require(!(min is Int && min.toInt() > max.toInt())) { "min must be less than max, min is $min, max is $max" }
        }

        /**
         * compare two `Number` of the same type.
         *
         * @param a   the first value.
         * @param b   the second value.
         * @param <T> The type of the number.
         * @return the value `0` if `d1` is numerically equal to `d2`; a value less than
         * `0` if `d1` is numerically less than `d2`; and a value greater than `0`
         * if `d1` is numerically greater than `d2`.
        </T> */
        fun <T : Number> compare(a: T, b: T): Int {
            if (a is Double) {
                return (a as Double).compareTo((b as Double))
            }
            if (a is Long) {
                return (a as Long).compareTo((b as Long))
            }
            if (a is Int) {
                return (a as Int).compareTo((b as Int))
            }
            throw IllegalArgumentException("Unsupported type " + a.javaClass)
        }
    }
}
